File: RewardCore/Sources/RewardCore/Engine/StreakCalculator.swift
import Foundation

/// 连击（连续达标天数）计算：从 asOf 当天往前数，连续每天净得分 > 0 的天数。
/// 当天净得分 ≤ 0 即连击为 0。
public enum StreakCalculator {

    public static func currentStreak(
        events: [ScoreEvent],
        asOf: Date,
        calendar: Calendar = .current
    ) -> Int {
        var netByDay: [Date: Int] = [:]
        for event in events where !event.isVoided {
            let dayStart = calendar.startOfDay(for: event.timestamp)
            netByDay[dayStart, default: 0] += event.points
        }

        var streak = 0
        var cursor = calendar.startOfDay(for: asOf)
        while let net = netByDay[cursor], net > 0 {
            streak += 1
            guard let previous = calendar.date(byAdding: .day, value: -1, to: cursor) else { break }
            cursor = previous
        }
        return streak
    }
}

File: web/src/domain/streakCalculator.ts
// StreakCalculator 移植：连续「净分 > 0」天数；净 0 / 负 / 缺天 / 全 voided 都断连。

import type { ScoreEvent } from './types'
import { addDays, startOfDay } from './dates'

export function currentStreak(events: ScoreEvent[], asOf: Date): number {
  const netByDay = new Map<number, number>()
  for (const e of events) {
    if (e.isVoided) continue
    const key = startOfDay(e.timestamp).getTime()
    netByDay.set(key, (netByDay.get(key) ?? 0) + e.points)
  }

  let streak = 0
  let cursor = startOfDay(asOf)
  while ((netByDay.get(cursor.getTime()) ?? 0) > 0 && netByDay.has(cursor.getTime())) {
    streak += 1
    // 再套一层 startOfDay：极端时区（如智利）本地午夜可能不存在，addDays 会落在 01:00，
    // 必须重归一化才能与 netByDay 的 key（各天 startOfDay）对得上
    cursor = startOfDay(addDays(cursor, -1))
  }
  return streak
}
